cmake/multi_image.cmake

```diff
@@ -23,125 +23,15 @@ if(IMAGE_NAME)
-endfunction()
-
-function(add_child_image_from_source)
-  # See 'add_child_image'
-  set(oneValueArgs NAME SOURCE_DIR DOMAIN BOARD)
-  cmake_parse_arguments(ACI "" "${oneValueArgs}" "" ${ARGN})
-
-  if (NOT ACI_NAME OR NOT ACI_SOURCE_DIR)
-    message(FATAL_ERROR "Missing parameter, required: NAME SOURCE_DIR")
-  endif()
-
-  # Pass information that the partition manager is enabled to Kconfig.
-  add_overlay_config(
-    ${ACI_NAME}
-    ${NRF_DIR}/subsys/partition_manager/partition_manager_enabled.conf
-    )
-
-  if (${ACI_NAME}_BOARD)
-    message(FATAL_ERROR
-      "${ACI_NAME}_BOARD set in outer scope. Will be ignored, use "
-      "`add_child_image(BOARD ${${ACI_NAME}_BOARD} ...)` for adding a child "
-      "image for specific board")
-  endif()
-
-  # Add the new partition manager domain if needed.
-  # The domain corresponds to the BOARD without the 'ns' suffix.
-  if (ACI_DOMAIN)
-    if ("${ACI_BOARD}" STREQUAL "")
-      message(FATAL_ERROR
-        "No board specified for domain '${ACI_DOMAIN}'. This configuration is "
-        "typically defined in ${BOARD_DIR}/Kconfig")
-    endif()
-
-    set(domain_parent ${${ACI_DOMAIN}_PM_DOMAIN_DYNAMIC_PARTITION})
-    if(DEFINED ${ACI_DOMAIN}_PM_DOMAIN_DYNAMIC_PARTITION
-       AND NOT "${domain_parent}" STREQUAL "${ACI_NAME}"
-    )
-      # A domain may only have one child image, which can then act as a parent
-      # to other images in the domain.
-      # As it is a cache variable we check it's content so that CMake re-run
-      # will pass the check as long as the child image hasn't changed.
-      message(FATAL_ERROR "A domain may only have a single child image."
-        "Current domain image is: ${domain_parent}, `${domain_parent}` is a "
-	"domain parent image, so you may add `${ACI_NAME}` as a child inside "
-	"`${domain_parent}`"
-      )
-    endif()
-    # This needs to be made globally available as it is used in other files.
-    set(${ACI_DOMAIN}_PM_DOMAIN_DYNAMIC_PARTITION ${ACI_NAME} CACHE INTERNAL "")
-
-    if (NOT (${ACI_DOMAIN} IN_LIST PM_DOMAINS))
-      list(APPEND PM_DOMAINS ${ACI_DOMAIN})
-      set_property(GLOBAL APPEND PROPERTY PM_DOMAINS ${ACI_DOMAIN})
-    endif()
-  elseif (NOT ACI_BOARD)
-    # No BOARD is given as argument, this triggers automatic conversion of
-    # *.ns board from parent image.
-    get_board_without_ns_suffix(${BOARD} ACI_BOARD)
-  endif()
-
-  if (NOT ACI_DOMAIN AND DOMAIN)
-    # If no domain is specified, a child image will inherit the domain of
-    # its parent.
-    set(ACI_DOMAIN ${DOMAIN})
-    set(inherited " (inherited)")
-  endif()
-
-  set(${ACI_NAME}_DOMAIN ${ACI_DOMAIN})
-  set(${ACI_NAME}_BOARD ${ACI_BOARD})
-
-  message("\n=== child image ${ACI_NAME} - ${ACI_DOMAIN}${inherited} begin ===")
+#
+# Function for gathering all CMake arguments relevant to Kconfig and Devicetree
+# for a child image. The output can be passed directly to a CMake call.
+#
+function(get_child_image_cmake_args child_image output_cmake_args)
```

Comment:
I understand your concern, it looks like this function is supposed to be called from outside this file. The reason I pulled this out to its own function was to make the code inside `add_child_image` more readable. In particular this section:

```
  message("\n=== child image ${ACI_NAME} - ${ACI_DOMAIN}${inherited} begin ===")

  if (PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME})
    # A complete set of configurations has been specified. Use this instead of
    # finding the cmake args the normal way.
    set(image_cmake_args "${PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME}}")
  else()
    get_child_image_cmake_args(${ACI_NAME} image_cmake_args)
  endif()
```
Where it is easy for the user to see that the configuration is either picked from the variable, or fetched from that function. If we inline the function here it is harder to understand what is going on.

As far as I see we have two options (feel free to add more if you see them):
1. Keep the function and perhaps point out that it is only intended for internal use (although in theory you _can_ use it from outside.
2. Inline the function, and live with the (IMO) slightly less readable option.

What option do you think is best? Or do you have another option?